org.osgi.test.common/src/main/java/org/osgi/test/common/annotation/WithService.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) OSGi Alliance (2020). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.osgi.test.common.annotation;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+@Inherited
+@Target({
+	ElementType.TYPE, ElementType.METHOD
+})
+@Retention(RUNTIME)
+@Documented
+@Repeatable(WithServices.class)
+public @interface WithService {
+
+	String ofMethod() default "";
+
+	Class<?> ofClass() default Object.class;
+
+	/**
+	 * The services of the Component. If not set Object.class will be used.
+	 *
+	 * @return The services.
+	 */
+	Class<?>[] value() default {
+		Object.class
+	};
+
+	ServiceProperty[] properties() default {
+		@ServiceProperty(key = ServiceProperty.NOT_SET)
```

Comment:
Why isn't the default just an empty array?

Reply:
Then it would generate. Empty hashmap. This will be used to update the config e.g. in configadmin and remove all existing configuration.

Reply:
Well, the code here processes the properties element. So it can decide that an empty array means no service properties, no?

Reply:
absolutly agree here.